### PR TITLE
Implementing HF Padding-Free and GraniteLM Support

### PR DESCRIPTION
**Pull request by aldopareja**
Updating the data collator for models with HF padding-free support, adding support for upcoming Granite HF model class, and updating flags/interface accordingly.

 -Mustafa

### CODE REVIEW

File: src/instructlab/training/token_dataset.py [hunk 1]
```diff
@@ -11,18 +11,25 @@
...
-        )
+        if "len" not in self.data.column_names:
+            self.lengths = np.array(
+                self.data.map(
+                    lambda x: {"len": len(x["input_ids"])},
+                    num_proc=8,
+                )["len"]
+            )
+        else:
+            self.lengths = np.array(self.data["len"])
```
Comment (RobotSail) [inline src/instructlab/training/token_dataset.py:32]: I'm a bit unfamiliar with setting up datasets like this - is there a reason we'd want to set this up as a numpy array first instead of just using `torch.tensor` directly?

File: src/instructlab/training/utils.py [hunk 2]
```diff
@@ -111,9 +111,18 @@ def listen(self):
...
 
 
-def make_collate_fn(pad_token_id, is_granite=False, max_batch_len=60000):
+def supports_flash_attention(device_id=0):
+    """Check if a GPU supports FlashAttention."""
+    major, minor = torch.cuda.get_device_capability(device_id)
+    # Check if the GPU architecture is Ampere (SM 8.x) or newer (SM 9.0)
+    is_sm8x = major == 8 and minor >= 0
+    is_sm90 = major == 9 and minor == 0
+    return is_sm8x or is_sm90
```
Comment (RobotSail) [inline src/instructlab/training/utils.py:120]: What happens if someone runs this on AMD, would this be able to handle that case as well?

File: src/instructlab/training/utils.py [hunk 3]
```diff
@@ -140,70 +149,108 @@ def pad_collate_fn(batch):
...
+                input_ids = []
+                labels = []
+                position_ids = []
+                total_len = 0
+                num_loss_counted_tokens = 0
+
+                for num_samples, item in enumerate(batch):
+                    item_len = len(item["input_ids"])
+                    if total_len + item_len > max_batch_len:
+                        break
```
Comment (RobotSail) [inline src/instructlab/training/utils.py:168]: Could this potentially drop samples if we're breaking after taking a sample from the batch without putting it back?

File: src/instructlab/training/utils.py (hunk 2 above)
Comment (Maxusmusti) [reply src/instructlab/training/utils.py:120]: No, I'll need to update this to handle it

File: src/instructlab/training/token_dataset.py (hunk 1 above)
Comment (aldopareja) [reply src/instructlab/training/token_dataset.py:28]: to get the mean() with a single command later on. Prefer numpy to make sure it's in CPU.

File: src/instructlab/training/utils.py (hunk 3 above)
Comment (aldopareja) [reply src/instructlab/training/utils.py:187]: yes but it doesn't matter, and also, with multipack it never goes beyond max batch len so this never happens.

File: src/instructlab/training/data_process.py [hunk 4]
```diff
@@ -347,9 +347,24 @@ def main(args: DataProcessArgs):
...
     # extract only labels and messages formatted into a new dataset
-    data_with_labels = data_with_labels.select_columns(["labels", "input_ids"])
+    data_with_labels = data_with_labels.map(
+        lambda x: {
+            "len": len(x["input_ids"]),
+        },
+        num_proc=NUM_PROC,
+    )
+    data_with_labels = data_with_labels.select_columns(["labels", "input_ids", "len"])
+    max_id = len(tokenizer) - 3
```
Comment (JamesKunstle) [inline src/instructlab/training/data_process.py:357]: what is the -3 for?

File: src/instructlab/training/data_process.py [hunk 5]
```diff
@@ -347,9 +347,24 @@ def main(args: DataProcessArgs):
...
+    final_valid_data = data_with_labels.filter(
+        lambda x: all(tk < max_id for tk in x["labels"]), num_proc=NUM_PROC
+    )
+    if len(final_valid_data) < len(data_with_labels):
+        dropped_samples = len(data_with_labels) - len(final_valid_data)
+        print(
+            f"\033[93mWarning: {dropped_samples} samples were dropped because they contained token IDs greater than or equal to {max_id}.\033[0m"
```
Comment (JamesKunstle) [inline src/instructlab/training/data_process.py:364]: can this be a problem ever? I imagine that any unaccounted-for tokens would be assigned to UNK but maybe the dropped tokens (from the above filtering step) should be logged just in case.

File: src/instructlab/training/main_ds.py [hunk 6]
```diff
@@ -102,15 +104,11 @@ def setup_model(args, tokenizer, train_loader, grad_accum):
         "torch_dtype": torch.bfloat16,
         "quantization_config": bnb_config,
     }
-    if not args.disable_flash_attn:
+    if flash_enabled:
         base_model_args["attn_implementation"] = "flash_attention_2"
```
Comment (JamesKunstle) [inline src/instructlab/training/main_ds.py:108]: is `flash_attention_2` going to be coming from the `FlashAttention` repo or is this something that will need to be built by core components?

File: src/instructlab/training/main_ds.py [hunk 7]
```diff
@@ -639,28 +645,26 @@ def run_training(torch_args: TorchrunArgs, train_args: TrainingArgs) -> None:
...
-                #
-                # An important reason for why #1 would be preferable is in the case of OpenShift/SELinux
-                # where the user has a defined place for new temporary data to be written.
-                data_output_path=train_args.data_output_dir,
-                model_path=train_args.model_path,
-                data_path=train_args.data_path,
-                max_seq_len=train_args.max_seq_len,
-                chat_tmpl_path=train_args.chat_tmpl_path,
-            )
+    check_valid_train_args(train_args)
```
Comment (JamesKunstle) [inline src/instructlab/training/main_ds.py:648]: I think this removes the optional data processing check on line 648 that @MichaelClifford added a while ago. This should be kept.

File: src/instructlab/training/utils.py [hunk 8]
```diff
@@ -40,6 +41,44 @@
...
+    # early validation logic here
+    if train_args.max_batch_len < train_args.max_seq_len:
+        raise ValueError(
+            f"the `max_batch_len` cannot be less than `max_seq_len`: {train_args.max_batch_len=} < {train_args.max_seq_len=}"
+        )
+
+    if os.path.exists(train_args.model_path):
+        if not os.path.isdir(train_args.model_path):
+            raise FileNotFoundError(
+                "Model path does not appear to be a dir, please validate or update the path"
```
Comment (JamesKunstle) [inline src/instructlab/training/utils.py:58]:
```suggestion
                "Model path does not appear to be a directory. Please make sure that you're passing a Hugging Face Transformers compatible directory checkpoint."
```

File: src/instructlab/training/utils.py [hunk 9]
```diff
@@ -40,6 +41,44 @@
...
+            )
+    else:
+        raise FileNotFoundError(
+            "Model Path cannot be found, please verify existence and permissions"
```
Comment (JamesKunstle) [inline src/instructlab/training/utils.py:62]:
```suggestion
            f"Provided path to model does not exist. Please make sure that you've passed a valid model and that it has appropriate permissions: {train_args.model_path}"
```

**Review by JamesKunstle — COMMENTED** [review]
Generally seems like a solid drop-in replacement with some good updates. First round of review only covers code- I'll run everything on AMD hardware as well.

File: src/instructlab/training/main_ds.py (hunk 7 above)
Comment (Maxusmusti) [reply src/instructlab/training/main_ds.py:648]: Oh good catch, let me fix this

File: src/instructlab/training/main_ds.py (hunk 6 above)
Comment (Maxusmusti) [reply src/instructlab/training/main_ds.py:108]: This is a model arg available during model instantiation (via transformers), if included it will import and utilize the flash attention package

File: src/instructlab/training/main_ds.py (hunk 7 above)
Comment (Maxusmusti) [reply src/instructlab/training/main_ds.py:648]: ok fixed

File: src/instructlab/training/utils.py (hunk 8 above)
Comment (Maxusmusti) [reply src/instructlab/training/utils.py]: updated

File: src/instructlab/training/utils.py (hunk 9 above)
Comment (Maxusmusti) [reply src/instructlab/training/utils.py]: updated

File: src/instructlab/training/data_process.py (hunk 4 above)
Comment (aldopareja) [reply src/instructlab/training/data_process.py:357]: because MASK and both pretrain tokens should not be in the tokens, those are special tokens added only for data processing purposes.

File: src/instructlab/training/data_process.py (hunk 5 above)
Comment (aldopareja) [reply src/instructlab/training/data_process.py:364]: yes, a dataset I was training on had the same string that we used for the MASK token, then the integer that would be mapping to would be higher than the integers the model would have and the training would break. This is only for the tokens that are added here but not in training.

File: src/instructlab/training/main_ds.py (hunk 6 above)
Comment (JamesKunstle) [reply src/instructlab/training/main_ds.py:108]: 👍

File: src/instructlab/training/data_process.py (hunk 4 above)
Comment (JamesKunstle) [reply src/instructlab/training/data_process.py:357]: Could you please add an explanation comment for that? Raw integers like that are called "magic numbers" and are usually managed by either naming a constant to replace it or commenting what it means.

Comment (Maxusmusti) [reply src/instructlab/training/data_process.py:357]: I'll do that

File: src/instructlab/training/data_process.py (hunk 5 above)
Comment (Maxusmusti) [reply src/instructlab/training/data_process.py:364]: I'll do that

Comment (Maxusmusti) [reply src/instructlab/training/data_process.py:366]: done

File: src/instructlab/training/data_process.py (hunk 4 above)
Comment (Maxusmusti) [reply src/instructlab/training/data_process.py:358]: done

**Comment by Maxusmusti** [comment]
@JamesKunstle added comments for the things you wanted

**Comment by JamesKunstle** [comment]
@Maxusmusti Much appreciated- testing on AMD now

**Comment by JamesKunstle** [comment]
Testing passes on AMD, losses between Dolomite and Llama are in parity.

**Review by JamesKunstle — APPROVED** [review]
Tests pass and loss curves look good between Dolomite and Llama padding-free on AMD. Feel good about merging this.

File: src/instructlab/training/data_process.py [hunk 10]
```diff
@@ -199,7 +199,7 @@ def print_masked_samples(data, tokenizer, is_pretrain, num_proc):
     def get_masked_and_orig_text(sample):
         labels = sample["labels"]
         input_ids = sample["input_ids"]
-        mask_id = get_sp_token(tokenizer, "<MASK>")[0]
+        mask_id = get_sp_token(tokenizer, "<|MASK|>")[0]
```
Comment (RobotSail) [inline src/instructlab/training/data_process.py:202]: Will this affect existing models? Or is this purely for training-time?

**Review by RobotSail — APPROVED** [review]
LGTM

File: src/instructlab/training/data_process.py (hunk 10 above)
Comment (Maxusmusti) [reply src/instructlab/training/data_process.py:202]: Yeah only relevant during training
